[MRG + 1] load_iris dataset: added return_X_y option

manu-chroma opened the pull request:
<!--
Thanks for contributing a pull request! Please ensure you have taken a look at
the contribution guidelines: https://github.com/scikit-learn/scikit-learn/blob/master/CONTRIBUTING.md#Contributing-Pull-Requests
-->
#### Reference Issue

<!-- Example: Fixes #1234 -->

Add return_tuple option to data loaders that return a Bunch https://github.com/scikit-learn/scikit-learn/issues/6670
#### What does this implement/fix? Explain your changes.
1. added return_X_y option in load_iris
2. Wrote tests for the same 
#### Any other comments?
1. Majority of the code is taken from https://github.com/scikit-learn/scikit-learn/pull/6704
2. Some improvements based on the discussion on the same PR 
3. First contribution to scikit-learn 

<!--Please be aware that we are a loose team of volunteers so patience is
necessary; assistance handling other issues is very welcome. We value
all user contributions, no matter how minor they are. If we are slow to
review, either the pull request needs some benchmarking, tinkering,
convincing, etc. or more likely the reviewers are simply busy. In either
case, we ask for your understanding during the review process.
For more information, see our FAQ on this topic:
http://scikit-learn.org/dev/faq.html#why-is-my-pull-request-not-getting-any-attention.

Thanks for contributing!
-->

manu-chroma committed:
load_iris dataset:added return_X_y option

nelson-liu commented on sklearn/datasets/base.py:
```diff
@@ -255,6 +255,12 @@ def load_iris():
+
     Returns
     -------
     data : Bunch
```
I think you should change this line to reflect how the function can return a Bunch or a tuple now.

manu-chroma replied:
That's precisely what this line defines. Don't you think ?
```
  If True, returns (data, target) instead of a Bunch object.
```

nelson-liu replied:
Well on line 266 in the diff it still says that it returns `data : Bunch`. Would it be apt to include information about it possibly returning (data, target) as well?

manu-chroma replied:
Got it thanks.

nelson-liu commented on sklearn/datasets/base.py:
```diff
@@ -255,6 +255,12 @@ def load_iris():
+     ----------
+    return_X_y : boolean, default=False.
+        If True, returns (data, target) instead of a Bunch object.
+        See below for more information about the `data` and `target` object
```
add a period at the end of this line?

manu-chroma replied:
Done. Also updated the return type description.

manu-chroma committed:
Updated return type description

nelson-liu commented on sklearn/datasets/base.py:
```diff
@@ -255,14 +255,25 @@ def load_iris():
+
     Returns
     -------
+    if return_X_y == false
```
i think you can actually omit this...i'm thinking of the notion that the "default" case is returning the `Bunch`, and you return the tuple `if return_X_y is True`

nelson-liu commented on sklearn/datasets/base.py:
```diff
@@ -255,14 +255,25 @@ def load_iris():
         full description of the dataset.
+    
+    if return_X_y == true
+    (data,target) : tuple
```
I think this would look better in the docs as  `(data, target) : tuple if return_X_y is True`...I can't think of any precedent for this in the code base; could you generate the docs by running `make html-noplot` in the `docs/` directory and post a screenshot of what it looks like?

manu-chroma replied:
Ok. I'll do it in a few hours.

manu-chroma replied:
Hey, I'm a little confused here. The methods and description doesn't seem to be updated after I ran the command.

Path of html page: ```file:///home/manu/github/scikit-learn/doc/_build/html/stable/modules/generated/sklearn.datasets.load_iris.html#sklearn.datasets.load_iris```

![image](https://cloud.githubusercontent.com/assets/12952080/16981428/32af687e-4e88-11e6-8b33-ab92e6bfb977.png)

manu-chroma committed:
improved return type description

jnothman commented on sklearn/datasets/base.py:
```diff
@@ -255,14 +255,25 @@ def load_iris():
         meaning of the features, and 'DESCR', the
         full description of the dataset.
+    
+    if return_X_y == true
```
numpydoc can't parse this line

manu-chroma replied:
I updated to this in the latest commit.
```
(data,target) : tuple if return_X_y == true
```
Still not parsing. 
Also, the parameter ```(return_X_y=False)``` is also not visible in the method definition.

maniteja123 replied:
Hi, in case even the method signature isn't changed, probably the docs aren't getting generated correctly. Can you please check that ?

nelson-liu replied:
the way i usually do this is hit the link at the bottom of the page "view this page source" and verify that the source is what is in my file.

manu-chroma replied:
After making the changes, I generated the docs by running ```make html-noplot``` in```doc/``` folder.
Path of html page: ```file:///home/manu/github/scikit-learn/doc/_build/html/stable/modules/generated/sklearn.datasets.load_iris.html#sklearn.datasets.load_iris```

What is it, I might be doing wrong ?

manu-chroma replied:
@nelson-liu Sure, I'll do that.

manu-chroma replied:
I'm afraid page source is not helping.
![image](https://cloud.githubusercontent.com/assets/12952080/16995161/77a09918-4ec8-11e6-9cf7-e4a29becaaf7.png)

Clicking on the source in line with the method signature points to https://github.com/scikit-learn/scikit-learn/blob/3f9494c/sklearn/datasets/base.py#L242
Which, I suppose, is the most recent commit I made. 

![image](https://cloud.githubusercontent.com/assets/12952080/16995267/074c831a-4ec9-11e6-9a29-5db80f3b7361.png)

nelson-liu replied:
i downloaded the PR myself and built the docs, and it looks like this (which seems correct, bar a few nitpicks) to me
<img width="1274" alt="screen shot 2016-07-20 at 11 17 10 am" src="https://cloud.githubusercontent.com/assets/7272031/16997764/a69ff8ec-4e6b-11e6-9a0a-02cd594bd96d.png">

nelson-liu commented on sklearn/datasets/base.py:
```diff
@@ -255,6 +255,12 @@ def load_iris():
+    return_X_y : boolean, default=False.
+        If True, returns (data, target) instead of a Bunch object.
+        See below for more information about the `data` and `target` object.
+
```
i checked this PR out locally and this line has whitespace, can you remove that?

nelson-liu commented on sklearn/datasets/base.py:
```diff
@@ -263,6 +269,9 @@ def load_iris():
         'target_names', the meaning of the labels, 'feature_names', the
         meaning of the features, and 'DESCR', the
         full description of the dataset.
+    
```
i checked this PR out locally and this line has whitespace, can you remove that?

nelson-liu commented on sklearn/datasets/base.py:
```diff
@@ -263,6 +269,9 @@ def load_iris():
         full description of the dataset.
+    
+    (data,target) : tuple if return_X_y == true
+
```
extraneous newline

manu-chroma replied:
Thanks alot. @jnothman did point this fact in the beginning but I failed to understood.
Fixed.

nelson-liu commented on sklearn/datasets/base.py:
```diff
@@ -263,6 +269,9 @@ def load_iris():
         meaning of the features, and 'DESCR', the
         full description of the dataset.
+    
+    (data,target) : tuple if return_X_y == true
```
"true" is generally capitalized

manu-chroma replied:
Hey, regenerated the docs. Still not working after the latest commit. Any clue ?

nelson-liu replied:
hmm not sure, I'm not super familiar with the intricacies of sphinx and the doc building process. what are your versions (run the snippet below) + sphinx version?
```
import platform; print(platform.platform())
import sys; print("Python", sys.version)
import numpy; print("NumPy", numpy.__version__)
import scipy; print("SciPy", scipy.__version__)
import sklearn; print("Scikit-Learn", sklearn.__version__)
```

manu-chroma replied:
```
Linux-4.4.0-31-generic-x86_64-with-Ubuntu-16.04-xenial
('Python', '2.7.12 (default, Jul  1 2016, 15:12:24) \n[GCC 5.4.0 20160609]')
('NumPy', '1.11.1')
('SciPy', '0.17.0')
('Scikit-Learn', '0.18.dev0')

Sphinx (sphinx-build) 1.3.6
```

nelson-liu replied:
try upgrading your sphinx (I'm on 1.4.5). Otherwise, our systems look pretty similar
```
Darwin-15.5.0-x86_64-i386-64bit
('Python', '2.7.12 (default, Jun 29 2016, 14:05:02) \n[GCC 4.2.1 Compatible Apple LLVM 7.3.0 (clang-703.0.31)]')
('NumPy', '1.11.1')
('SciPy', '0.17.1')
('Scikit-Learn', '0.18.dev0')
```

manu-chroma replied:
Updated and added an extra line.

manu-chroma committed:
Removed extra line

nelson-liu commented on sklearn/datasets/base.py:
```diff
@@ -263,6 +269,7 @@ def load_iris():
         'target_names', the meaning of the labels, 'feature_names', the
         meaning of the features, and 'DESCR', the
         full description of the dataset.
+    (data,target) : tuple if return_X_y == True
```
sorry, i mean that there should be a newline above the new return type but not one below it.

manu-chroma committed:
Added extra line

manu-chroma committed:
Added test sentence.

manu-chroma commented on sklearn/datasets/base.py:
```diff
@@ -255,14 +255,22 @@ def load_iris():
         'target_names', the meaning of the labels, 'feature_names', the
         meaning of the features, and 'DESCR', the
-        full description of the dataset.
+        full description of the dataset. This is test text.
```
@nelson-liu Hey, I added this test sentence and it as well didn't show up on re-building the docs.

I've also tried updating ```Sphinx to 1.4.5``` 
Really confused here.

nelson-liu replied:
how are you building? you should be navigating to the `doc/` subdirectory and then running `make html-noplot` there. the build output should show up under at `doc/_build/html/stable/modules/generated/sklearn.datasets.load_iris.html`

manu-chroma replied:
Yep. That's what I'm doing. 

![image](https://cloud.githubusercontent.com/assets/12952080/17032804/b54a2d3a-4f98-11e6-823f-1299516cbf59.png)


![image](https://cloud.githubusercontent.com/assets/12952080/17032777/94516238-4f98-11e6-9055-e07f9ee0b2ed.png)

amueller replied:
try "make clean" beforehand

nelson-liu replied:
whoops this totally fell off my radar, thanks for the suggestion Andy

manu-chroma replied:
@amueller @nelson-liu No change. :disappointed:

nelson-liu replied:
try uninstalling and reinstalling? Make sure you follow the instructions [here](http://scikit-learn.org/stable/developers/contributing.html#retrieving-the-latest-code).

manu-chroma replied:
hey, Quick question:
Initially, I installed using ```make``` rather than using ```python setup.py build_ext --inplace```

That shouldn't make any difference. Right ?

nelson-liu replied:
I'm not too sure...try using `python setup,py develop`

amueller replied:
I think I know the problem ;) sphinx uses your system installation of scikit-learn. Make sure that if you run python inside the doc folder and import scikit-learn, your dev version is imported.

manu-chroma replied:
Hey @amueller, thanks for your help.

Yes, my dev version is imported when I try doing that. Also, I ran ```pip list``` and only the dev version installation is shown in the list. 

![image](https://cloud.githubusercontent.com/assets/12952080/17220420/c3a6c2c0-550c-11e6-9e41-4850953186a6.png)

nelson-liu replied:
oh TIL, i never knew that sphinx used the system install vs the install that it's run from. hopefully that works @manu-chroma

nelson-liu replied:
does running `python -c "import sklearn; print sklearn.__file__"` give you the path `~/github/scikit-learn/sklearn/__init__.pyc`?

amueller replied:
Well, it uses ``import sklearn``, so it uses what ever that yields. ;)

manu-chroma replied:
It doesn't!
I'm not able to understand why it isn't pointing to the path you pointed out.

![image](https://cloud.githubusercontent.com/assets/12952080/17220648/c8219b3a-550d-11e6-902c-480ddbe8814d.png)

nelson-liu replied:
You probably pip installed the github dev repo at some point in time ; remove that / all scikit learn installs (I'm assuming you only need one) and then reinstall with the development instructions above

nelson-liu replied:
this is what `pip list` shows on my computer: 
<img width="497" alt="screen shot 2016-07-28 at 9 43 40 am" src="https://cloud.githubusercontent.com/assets/7272031/17221254/f737f25e-54a7-11e6-936c-046f079debfc.png">

manu-chroma replied:
Reinstalled! :)
Building docs now
![image](https://cloud.githubusercontent.com/assets/12952080/17221398/62fee448-5511-11e6-97c1-ca78c76a3c46.png)

manu-chroma replied:
This makes me really happy.
Thanks alot @nelson-liu for your constant feedback and @amueller for your insights. :grinning: 

![image](https://cloud.githubusercontent.com/assets/12952080/17221989/c0e068f0-5513-11e6-98a0-91514174e762.png)

manu-chroma committed:
Remove sample text

nelson-liu commented on sklearn/datasets/base.py:
```diff
@@ -264,6 +270,8 @@ def load_iris():
         meaning of the features, and 'DESCR', the
         full description of the dataset.
 
+    (data,target) : tuple if return_X_y == True
```
i prefer "...tuple if return_X_y is True", the `==` looks a bit odd to me. Additionally, i think it'd be nice to add to add two backticks (`) to the start and front of `return_X_y`.

amueller replied:
+1. Also, space after ``,`` ;)

amueller commented on sklearn/datasets/base.py:
```diff
@@ -255,6 +255,12 @@ def load_iris():
+    Parameters
+     ----------
+    return_X_y : boolean, default=False.
+        If True, returns (data, target) instead of a Bunch object.
```
double backticks around ``(data, target)``

amueller commented on sklearn/datasets/base.py:
```diff
@@ -255,6 +255,12 @@ def load_iris():
+     ----------
+    return_X_y : boolean, default=False.
+        If True, returns (data, target) instead of a Bunch object.
+        See below for more information about the `data` and `target` object.
```
Double backticks

amueller commented on sklearn/datasets/tests/test_base.py:
```diff
@@ -180,6 +181,13 @@ def test_load_iris():
+    #test return_X_y option
+    X_y_tuple = load_iris(return_X_y=True)
+    bunch = load_iris()
+    assert_true(isinstance(X_y_tuple,tuple))
```
pep8: space after comma

amueller commented:
lgtm apart from comments

amueller commented:
I restarted CI

manu-chroma committed:
Fixes

manu-chroma commented:
@amueller I did the fixes. Do I need to squash commits or do anything else here ?

amueller commented on sklearn/datasets/base.py:
```diff
@@ -264,6 +270,8 @@ def load_iris():
         meaning of the features, and 'DESCR', the
         full description of the dataset.
 
+    (data,target) : tuple if ``return_X_y`` is True
```
pep8?

amueller commented on sklearn/datasets/tests/test_base.py:
```diff
@@ -180,6 +181,13 @@ def test_load_iris():
     assert_equal(res.target_names.size, 3)
     assert_true(res.DESCR)
 
+    #test return_X_y option
```
pep8: single space after ``#``

amueller commented:
lgtm apart from style comments. No, nothing else to do (except fix these)

manu-chroma committed:
pep8

amueller commented:
@nelson-liu +1 ?

nelson-liu commented:
still looking, give me a second

nelson-liu commented:
yup LGTM thanks @manu-chroma

nelson-liu commented:
does the warrant a what's new? or should there be one after this is implemented in all the loaders.

nelson-liu commented:
Also perhaps a versionadded?

amueller commented:
good catch. definitely a versionadded for the keyword. We can also add a whatsnew, and edit that in the future when we add this functionality to the other functions.

nelson-liu commented:
@manu-chroma does that sound doable? You would have to add an entry in [`whats_new.rst`](https://github.com/scikit-learn/scikit-learn/blob/master/doc/whats_new.rst), and the appropriate "versionadded" tags below your new parameter (see [this as an example](https://github.com/scikit-learn/scikit-learn/pull/7093/files))

manu-chroma commented:
@nelson-liu Sure, I can do that. 

I was thinking of updating the ```whats_new.rst``` after I have implemented this for all loaders, which I suppose wouldn't take more than a few days. Thoughts ?

nelson-liu commented:
yeah, that's one possibility but i think it would be a bit better to create an entry now, and then edit it as the functionality for more loaders is added with their associated PR's.

manu-chroma commented:
Okay, I'll update the PR in a few hours. I suppose this change should come under ```Enhancements``` in ```whats_new.rst``` ?

nelson-liu commented:
yup, that's a good assumption.

manu-chroma committed:
Added version tag

manu-chroma commented:
Hey, I updated my branch before editing ```whats_new.rst```

After that I added the version tag (https://github.com/scikit-learn/scikit-learn/pull/7049/commits/f4d6168886ae648ac94f5fbd90b2f842116753fa) and tried running ```sudo make html-noplot```
It failed. Here's the error log. Error is because of a different file. 

``` bash
manu@hp:~/github/scikit-learn/doc$ sudo make html-noplot 
sphinx-build -D plot_gallery=0 -b html -d _build/doctrees   . _build/html/stable
Running Sphinx v1.4.5
WARNING: sphinx.ext.pngmath has been deprecated. Please use sphinx.ext.imgmath instead.
loading pickled environment... not yet created
/home/manu/github/scikit-learn/sklearn/cross_validation.py:43: DeprecationWarning: This module was deprecated in version 0.18 in favor of the model_selection module into which all the refactored classes and functions are moved. Also note that the interface of the new CV iterators are different from that of this module. This module will be removed in 0.20.
  "This module will be removed in 0.20.", DeprecationWarning)
/home/manu/github/scikit-learn/sklearn/learning_curve.py:23: DeprecationWarning: This module was deprecated in version 0.18 in favor of the model_selection module into which all the functions are moved. This module will be removed in 0.20
  DeprecationWarning)

Exception occurred:
  File "/home/manu/github/scikit-learn/sklearn/tree/tree.py", line 61, in <module>
    "mae": _criterion.MAE}
AttributeError: 'module' object has no attribute 'MAE'
The full traceback has been saved in /tmp/sphinx-err-0pnBuo.log, if you want to report the issue to the developers.
Please also report this if it was a user error, so that a better error message can be provided next time.
A bug report can be filed in the tracker at <https://github.com/sphinx-doc/sphinx/issues>. Thanks!
Makefile:47: recipe for target 'html-noplot' failed
make: *** [html-noplot] Error 1
```

manu-chroma committed:
added entry in whats_new

manu-chroma commented:
Yes, that was the case. Really sorry for forgetting the standard procedure.

I've updated ```whats_new``` and added version tag. Thoughts ?

amueller commented on sklearn/datasets/base.py:
```diff
@@ -258,6 +258,15 @@ def load_iris():
     Read more in the :ref:`User Guide <datasets>`.
 
+    Parameters
+     ----------
```
There is an extra space here. Sphinx will complain.

manu-chroma replied:
Fixed. :+1:

manu-chroma committed:
fixed extra space

amueller commented:
thanks :)

nelson-liu commented on sklearn/datasets/base.py:
```diff
@@ -258,6 +258,15 @@ def load_iris():
+    Parameters
+    ----------
+    
+        .. versionadded:: 0.18
```
sorry i'm late to the party, shouldn't the `versionadded` be under `return_X_y`? and do we need a `versionadded` under the new return type (say around line 280)?

amueller replied:
yes. Do you want to do a PR to fix it? (sorry I'm swamped)

nelson-liu replied:
sure ill do it / fix the other comments i had on this PR @amueller @manu-chroma

manu-chroma replied:
Sorry for replying late.

I intend on sending the next PR this weekend (adding return_X_y option to load_breast_cancer dataset)
Should I take care of this ?

nelson-liu replied:
that's fine, i made a fix at #7138 . Just make sure to do it for next time :)

manu-chroma replied:
Sorry and I'll keep that in mind.

nelson-liu commented on doc/whats_new.rst:
```diff
@@ -230,6 +230,9 @@ Enhancements
      (`#6846 <https://github.com/scikit-learn/scikit-learn/pull/6846>`_)
      By `Sebastian Säger`_ and `YenChen Lin`_.
 
+   - Added new return type ``(data, target)`` : tuple option to :func:`load_iris` dataset. (`#7049 <https://github.com/scikit-learn/scikit-learn/pull/7049>`_) 
```
maybe split this into two lines, this line is a lot longer than the other ones.

manu-chroma replied:
I can fix this in the next PR (adding ```return_X_y``` option to load_breast_cancer dataset)
Also, the position of the ```versionadded``` tag. What do you think ?

nelson-liu replied:
sure sounds good. thanks!

nelson-liu commented on doc/whats_new.rst:
```diff
@@ -230,6 +230,9 @@ Enhancements
      By `Sebastian Säger`_ and `YenChen Lin`_.
 
+   - Added new return type ``(data, target)`` : tuple option to :func:`load_iris` dataset. (`#7049 <https://github.com/scikit-learn/scikit-learn/pull/7049>`_) 
+     By `Manvendra Singh`_ and `Nelson Liu`_.   
```
I shouldn't have any attribution on this, I didn't write any code / this is your contribution. thanks, though 😉

manu-chroma replied:
Your code review resulted in major changes in the initial PR . I think you had a decent part to play in this contribution. :smile:

nelson-liu replied:
nah, that's the job of a reviewer. thanks for the consideration though.
